spylib/utils/webhook.py

```diff
@@ -0,0 +1,91 @@
+from enum import Enum
+from typing import List, Optional, Union
+
+from pydantic import BaseModel
+
+from spylib.utils.misc import snake_to_camel_case
+
+
+class WebhookTopic(Enum):
+    ORDERS_CREATE = 'ORDERS_CREATE'
+
+
+class WebhookResponse(BaseModel):
+    id: str
+
+
+class WebhookSubscriptionCreate(Enum):
+    HTTP = 'webhookSubscriptionCreate'
+    EVENT_BRIDGE = 'eventBridgeWebhookSubscriptionCreate'
+    PUB_SUB = 'pubSubWebhookSubscriptionCreate'
+
+
+class WebhookSubscriptionInput(Enum):
+    HTTP = 'WebhookSubscriptionInput!'
+    EVENT_BRIDGE = 'EventBridgeWebhookSubscriptionInput!'
+    PUB_SUB = 'PubSubWebhookSubscriptionInput!'
+
+
+def generate_query(
+    webhook_subscription_create: WebhookSubscriptionCreate,
+    webhook_subscription_input: WebhookSubscriptionInput,
+) -> str:
+    """Generate the webhook creation graphql query"""
+    operation = webhook_subscription_create.value
+    input_type = webhook_subscription_input.value
+
+    query = f'''
+    mutation {operation}($topic: WebhookSubscriptionTopic!,
+                                        $webhookSubscription: {input_type}) {{
+        {operation}(topic: $topic, webhookSubscription: $webhookSubscription) {{
+            webhookSubscription {{
+            id
+            topic
+            format
+            endpoint {{
+                __typename
+                ... on WebhookHttpEndpoint {{
+                callbackUrl
+                }}
+                ... on WebhookEventBridgeEndpoint {{
+                arn
+                }}
+                ... on WebhookPubSubEndpoint {{
+                pubSubProject
+                pubSubTopic
+                }}
+              }}
+            }}
+            userErrors {{
+                field
+                message
+              }}
+            }}
+        }}
+    '''
+    return query
+
+
+def generate_variables(
+    topic: Union[WebhookTopic, str],
+    include_fields: Optional[List[str]] = None,
+    metafield_namespaces: Optional[List[str]] = None,
+    private_metafield_namespaces: Optional[List[str]] = None,
+    **endpoint_kwargs,
+) -> dict:
+    """Generate the webhook creation variables"""
+    endpoint_kwargs_camel_case_name = {}
+    for key, value in endpoint_kwargs.items():
+        endpoint_kwargs_camel_case_name[snake_to_camel_case(text=key)] = value
```

Comment:
I don't love this pattern. I assume this is done to reduce duplicate code however it comes at the cost of typing & being explicit. I would prefer something like the following even if it is longer:

```
def generate_shared_variables(
    topic: Union[WebhookTopic, str],
    include_fields: Optional[List[str]] = None,
    metafield_namespaces: Optional[List[str]] = None,
    private_metafield_namespaces: Optional[List[str]] = None,
):
    pass
  

# Do this in the method where is it needed
def generate_http_variables(..., url: str):
    generate_shared_variables()
    # merge vars here and return
```

Unfortunately [ParamSpec](https://docs.python.org/3/library/typing.html#typing.ParamSpec) is a `3.10` feature but one day!

Reply:
Thanks Quinn. I will change to the code to load the qgl like Ant suggested as well.

Reply:
Good to know about the [ParamSpec](https://docs.python.org/3/library/typing.html#typing.ParamSpec).
Yea, the intention was to stay DRY (do not repeat yourself), avoid copy and paste.  Thanks for pointing out the typing issue, @qw-in . I thought about making the input with pydantic basemodel. what would you think of that
```
class Webhook(BaseModel):  
    topic: Union[WebhookTopic, str] 
    includeFields: Optional[List[str]] = None
    metafieldNamespaces: Optional[List[str]] = None,
    privateMetafieldNamespaces: Optional[List[str]] = None

class HTTPWebhook(Webhook):
    callbackUrl: str

class EventBridgeWebhook(Webhook):
    arn: str

class PubSubWebhook(Webhook):
    ...



```